File: 74.search-a-2d-matrix.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution74 {
    // Treat this as a binary search with the root of the top right position.
    public bool SearchMatrix(int[][] matrix, int target) 
    {
        int rows = matrix.Length;
        if(rows == 0) return false;
        int columns = matrix[0].Length;
        if(columns ==0) return false;

        return BinarySearchMatrix(matrix, target, 0, columns -1 , rows, columns);
    }

    public bool BinarySearchMatrix(int[][] matrix, int target, int i, int j, int rows, int columns)
    {
        // return false if outside of the boundary
        if(i <0 || i >= rows || j < 0 || j >= columns)
        {
            return false;
        }

        if(matrix[i][j] == target)
        {
            return true;
        }
        else if(matrix[i][j] > target)
        {
            return BinarySearchMatrix(matrix, target, i, j - 1, rows, columns);
        }
        else
        {
            return BinarySearchMatrix(matrix, target, i + 1, j, rows, columns);
        }
    }
}

File: 74_search-a-2d-matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # Quick sort (Binary Search)
        m = len(matrix)
        if m == 0: return False
        n = len(matrix[0])
        if n == 0: return False

        return self.binarySearchMatrics(matrix, target, m, n, 0, n-1)

    def binarySearchMatrics(self, matrix, target, m, n, i, j):
        if i < 0 or i>=m or j <0 or j>= n: return False
        
        if matrix[i][j] == target: return True

        if matrix[i][j] > target:
            return self.binarySearchMatrics(matrix, target,  m, n, i, j -1)
        else:
            return self.binarySearchMatrics(matrix, target, m, n, i+1, j)

instance = Solution()
print(instance.searchMatrix([[1,3,5,7],[10,11,16,20],[23,30,34,60]], 3))